### PR TITLE
add mars redbank

### PR DESCRIPTION
**Pull request by g1nt0ki**
https://osmosis.marsprotocol.io/redbank

### CODE REVIEW

**Comment by llamatester** [comment]
The adapter at projects/mars-redbank exports TVL: 
        
 
 
    osmosis                   3.15 M
    osmosis-borrowed          1.58 M
    borrowed                  1.58 M
    
    total                    3.15 M

**Comment by larry0x** [comment]
@g1nt0ki Nice refactor using defillama-sdk, the code seems much cleaner.

1) My only concern is overflowing - the biggest number that nodejs supports is (2^53-1) which is just under 1e16, but in our case numbers can easily go above 1e18. Does the sdk use big numbers under the hood?

2) In terms of listing, I prefer to let it continue from the old Terra Classic Mars protocol, so that we see the full history of it going to zero and them reborn, instead of creating a new listing.

File: projects/mars-redbank/index.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+const chain = 'osmosis'
+const contract = 'osmo1c3ljch9dfw5kf52nfwpxd2zmj2ese7agnx0p9tenkrryasrle5sqf3ftpg'
+
+async function queryRedBankTvl() {
+  const res = await queryContract({ contract, chain: 'osmosis', data: { markets: {} } })
+
+  const tvl = {};
+  const borrowed = {};
+  res.forEach(i => {
+    sdk.util.sumSingleBalance(tvl, i.denom, i.collateral_total_scaled / 1e6 - i.debt_total_scaled / 1e6)
```
Comment (larry0x) [inline projects/mars-redbank/index.js:15]: This should be:

```suggestion
    sdk.util.sumSingleBalance(tvl, i.denom, i.collateral_total_scaled * i.liquidity_index / 1e6 - i.debt_total_scaled * i.debt_index / 1e6)
```

File: projects/mars-redbank/index.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  })
+
+  return { tvl: transformBalances(chain, tvl), borrowed: transformBalances(chain, borrowed) };
+}
+
+module.exports = {
+  timetravel: false,
+  methodology: "We query Mars protocol smart contracts to get the amount of assets deposited and borrowed, then use CoinGecko to price the assets in USD.",
+  osmosis: {
+    tvl: sumTokensExport({ owner: contract }),
```
Comment (larry0x) [inline projects/mars-redbank/index.js:26]: If the contract works as intended, this should always be equal to `(await queryRedBankTvl()).tvl`, so we can use either one here.

**Comment by g1nt0ki** [comment]
> @g1nt0ki Nice refactor using defillama-sdk, the code seems much cleaner.
> 
>     1. My only concern is overflowing - the biggest number that nodejs supports is (2^53-1) which is just under 1e16, but in our case numbers can easily go above 1e18. Does the sdk use big numbers under the hood?
> 
>     2. In terms of listing, I prefer to let it continue from the old Terra Classic Mars protocol, so that we see the full history of it going to zero and them reborn, instead of creating a new listing.

okay, I updated existing listing: https://github.com/DefiLlama/DefiLlama-Adapters/commit/78dd286010c7e34f0f79264df042c3ed767d1cde 

2. Hmm, I somehow have not run into this issue, wondering why? usually the value passed (to `sumSingleBalance` ) is in String format, and internally it uses BigNumber, I rarely pass in number format like this, will look into this.

Okay, just checked, it supports upto `1e308` 

> Number.MAX_VALUE
1.7976931348623157e+308
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_VALUE

File: projects/mars-redbank/index.js (hunk 2 above)
Comment (g1nt0ki) [reply projects/mars-redbank/index.js:26]: commented out tvl calculation in the other place, prefer query token balances for tvl.

**Comment by larry0x** [comment]
```
> Number.MAX_SAFE_INTEGER
9007199254740991
> Number.MAX_VALUE
1.7976931348623157e+308
```

Not sure what's the difference, but as long as we're not having overflowing in practice it's fine

**Comment by g1nt0ki** [comment]
> ```
> > Number.MAX_SAFE_INTEGER
> 9007199254740991
> > Number.MAX_VALUE
> 1.7976931348623157e+308
> ```
> 
> Not sure what's the difference, but as long as we're not having overflowing in practice it's fine

I am guessing that the first is an integer and second is float/double

**Pull request closed without merging**